### ISSUE TITLE
Well-known type JSON round-trip problem (possibly just C#)

### ISSUE THREAD

**Issue by jhickson** [google/protobuf#1509]
Hi,

[This might relate to just C# and might also just be my misunderstanding of the purpose of well-known types.]

If I have an instance of a well-known type (let's say BoolValue) that I write to JSON using `JsonFormatter.Default.Format()` in C# then the JSON generated is simply the wrapped value (e.g. `true`) rather than `{"value":true}` as I'd expect if I were using a well-known type at top-level.  Clearly the plain value string is not valid JSON and so I can't then parse it back to the well-known type.

Is this an issue in the code or are well-known types not meant to be used at the top-level?

Regards,

Jos

**Comment by xfxyjwf** [comment]
A JSON value can be a string in double quotes, or a number, or true or false or null, or an object or an array. The serialized format of a well-known type may not be a JSON object, but it's still valid JSON value. I think parsers should be able to handle that (e.g., be able to parse `true` as a BoolValue).

**Comment by jhickson** [comment]
Hi,

OK, I was under the misapprehension that top-level JSON needed to be an object.

After further investigation I think this is a bug in `JsonParser` that affects just `BoolValue`.  The test below fails:

```
[Test]
public void WellKnownTypeBoolValueJsonRoundTrip()
{
    var value = new BoolValue {Value = true};
    Assert.That(value, Is.EqualTo(JsonParser.Default.Parse<BoolValue>(JsonFormatter.Default.Format(value))));
}
```

but the test below for `Int32Value` passes

```
[Test]
public void WellKnownTypeInt32ValueJsonRoundTrip()
{
    var value = new Int32Value {Value = 23};
    Assert.That(value, Is.EqualTo(JsonParser.Default.Parse<Int32Value>(JsonFormatter.Default.Format(value))));
}
```

The problem would appear to be a lack of an entry for `BoolValue` in `JsonParser.WellKnownTypeHandlers`.

**Comment by jhickson** [comment]
Closing as fixed in #1517.
